frame/transaction-storage/src/lib.rs

```diff
@@ -0,0 +1,364 @@
+		/// Applies same fees as `store`.
+		/// # <weight>
+		/// - Constant with a single DB read.
+		/// # </weight>
+		#[pallet::weight(T::WeightInfo::renew())]
+		pub(super) fn renew(
+			origin: OriginFor<T>,
+			block: T::BlockNumber,
+			index: u32,
+		) -> DispatchResultWithPostInfo {
+			let info = <TransactionRoots<T>>::get(block, index).ok_or(Error::<T>::RenewedNotFound)?;
+			Self::apply_fee(origin, info.size)?;
+
+			let extrinsic_index = <frame_system::Pallet<T>>::extrinsic_index().unwrap();
+			sp_io::transaction_index::renew(extrinsic_index, info.content_hash.into());
+
+			let counter = <Counter<T>>::mutate(|c| { *c += 1; *c - 1 });
+			let block_num = <frame_system::Pallet<T>>::block_number();
+			<TransactionRoots<T>>::insert(block_num, counter, info);
+			Self::deposit_event(Event::Renewed(counter));
+			Ok(().into())
+		}
+
+		/// Check storage proof for block number `block_number() - StoragePeriod`.
+		/// If such block does not exist the proof is expected to be `None`.
+		/// # <weight>
+		/// - Linear w.r.t the number of indexed transactions in the proved block for random probing.
+		/// There's a DB read for each transaction.
+		/// Here we assume a maximum of 100 probed transactions.
+		/// # </weight>
+		#[pallet::weight((T::WeightInfo::check_proof_max(), DispatchClass::Mandatory))]
+		pub(super) fn check_proof(
+			origin: OriginFor<T>,
+			proof: Option<TransactionStorageProof>,
+		) -> DispatchResultWithPostInfo {
+			ensure_none(origin)?;
+			let number = <frame_system::Pallet<T>>::block_number();
+			let period = <StoragePeriod<T>>::get();
+			let target_number = number.saturating_sub(period);
+			if target_number.is_zero() {
+				ensure!(proof.is_none(), Error::<T>::UnexpectedProof);
+				return Ok(().into())
+			}
+			let transaction_count = <TransactionCount<T>>::get(target_number);
+			let mut total_chunks = 0;
+			for t in 0 .. transaction_count {
+				match <TransactionRoots<T>>::get(target_number, t) {
+					Some(info) => total_chunks +=
+						((info.size as u64 + CHUNK_SIZE as u64 - 1) / CHUNK_SIZE as u64) as u32,
+					None => break,
+				}
+			}
+			if total_chunks == 0 {
+				ensure!(proof.is_none(), Error::<T>::UnexpectedProof);
+				return Ok(().into())
+			}
+			let proof = proof.ok_or_else(|| Error::<T>::MissingProof)?;
+			let parent_hash = <frame_system::Pallet<T>>::parent_hash();
+			let selected_chunk_index = random_chunk(parent_hash.as_ref(), total_chunks);
+			total_chunks = 0;
+			let mut t = 0;
+			let (info, chunk_index) = loop {
+				match <TransactionRoots<T>>::get(target_number, t) {
+					Some(info) => {
+						let chunks = ((info.size as u64 + CHUNK_SIZE as u64 - 1) / CHUNK_SIZE as u64) as u32;
+						if total_chunks + chunks >= selected_chunk_index {
+							break (info, selected_chunk_index - total_chunks)
+						}
+						total_chunks += chunks;
+					},
+					None => Err(Error::<T>::MissingStateData)?,
+				}
+				t += 1;
+			};
+
+			ensure!(
+				sp_io::trie::blake2_256_verify_proof(
```

Comment:
So the purpose is to ensure users keep a copy of the storage proof for at least the storage period (somehow similar to having it in state but only by random sampling)?
If it is the case, then we need to write something in state: eg hash(previousblock ++ proof.chunk) so skipping will be a bit more difficult (need to trust a hash from state that we don't know).
(here just need to recompile the wasm to skip the check)

Reply:
Not sure what you mean here. Even if you recompile your node code, the rest of the network will still reject the block that does not contain the proof. The `check_proof` inherent is mandatory.

Reply:
I think I misunderstood how this work: the 'proof' parameter will be included in the block? (I was thinking every client would fetch their own storage for it).
So this will motivate only block producer to keep this content.

Reply:
(the scenario I was thinking of would have been implemented with a host function in fact (not an inherent)): calls to host function to access the data then hash it with prev block hash, so all network need the chunk sample (or fetch it somehow, which makes it a middleground between storage and speed and should probably need to query more than one chunk) and then check block from it.
What prevent user to skip this check is the inhability to build the next state.

Clearly, at this point I just do not really understand the purpose of this check, my hypothesis is checking that block producer keep access to this content.

Reply:
The chunk could be added to the proof the same way trie nodes are added, but that is certainly more impacting.